src/main/java/net/citizensnpcs/EventListen.java

```diff
@@ -423,34 +426,66 @@ public void onWorldUnload(WorldUnloadEvent event) {
+    public void onPlayerMove(final PlayerMoveEvent event) {
+
+        SkinUpdateTracker updateTracker = skinUpdateTrackers.get(event.getPlayer().getUniqueId());
+        if (updateTracker == null)
+            return;
+
+        if (!updateTracker.shouldUpdate(event.getPlayer()))
+            return;
+
+        updateTracker.reset(event.getPlayer());
+        recalculatePlayer(event.getPlayer(), 10, false);
+    }
+
+    public void recalculatePlayer(final Player player, long delay, final boolean isInitial) {
+
+        if (isInitial) {
+            skinUpdateTrackers.put(player.getUniqueId(), new SkinUpdateTracker(player));
+        }
+
         new BukkitRunnable() {
+
             @Override
             public void run() {
-                final List<EntityPlayer> nearbyNPCs = new ArrayList<EntityPlayer>();
-                for (NPC npc : getAllNPCs()) {
-                    Entity npcEntity = npc.getEntity();
-                    if (npcEntity instanceof Player && player.canSee((Player) npcEntity)
-                            && player.getWorld().equals(npcEntity.getWorld())
-                            && player.getLocation().distanceSquared(npcEntity.getLocation()) < 100 * 100) {
-                        nearbyNPCs.add(((CraftPlayer) npcEntity).getHandle());
-                    }
+
+                List<EntitySkinnable> nearbyNPCs = getNearbySkinnableNPCs(player);
+                for (EntitySkinnable npc : nearbyNPCs) {
+                     npc.getSkinTracker().updateViewer(player);
+                }
+
+                if (!nearbyNPCs.isEmpty() && isInitial) {
+                    // one more time to help when resource pack load times
+                    // prevent immediate skin loading
+                    recalculatePlayer(player, 40, false);
                 }
-                new BukkitRunnable() {
-                    @Override
-                    public void run() {
-                        sendToPlayer(player, nearbyNPCs);
-                    }
-                }.runTaskLater(CitizensAPI.getPlugin(), 30);
-                new BukkitRunnable() {
-                    @Override
-                    public void run() {
-                        sendToPlayer(player, nearbyNPCs);
-                    }
-                }.runTaskLater(CitizensAPI.getPlugin(), 70);
             }
-        }.runTaskLater(CitizensAPI.getPlugin(), 10);
+        }.runTaskLater(CitizensAPI.getPlugin(), delay);
+    }
+
+    private List<EntitySkinnable> getNearbySkinnableNPCs(Player player) {
+
+        List<EntitySkinnable> results = new ArrayList<EntitySkinnable>();
 
+        for (NPC npc : getAllNPCs()) {
+
+            Entity npcEntity = npc.getEntity();
+            if (npcEntity instanceof Player
+                    && player.canSee((Player) npcEntity)
+                    && player.getWorld().equals(npcEntity.getWorld())
+                    && player.getLocation().distanceSquared(npcEntity.getLocation()) < 100 * 100) {
```

Comment:
I'd prefer if you could use the cached Location methods (npc.getStoredLocation() and player.getLocation(Location)) to avoid object churn.